packages/grafana-data/src/transformations/transformers/insertValues.ts

```diff
@@ -0,0 +1,141 @@
+  threshold?: number;
+  // should refField's values be scanned to find minimum interval?
+  guessInterval?: boolean;
+}
+
+export const insertValuesTransformer: DataTransformerInfo<InsertValuesTransformerOptions> = {
+  id: DataTransformerID.insertValues,
+  name: 'Insert Values',
+  description: 'Can be used to insert null values between adjacent points spanning a minimum time threshold',
+  defaultOptions: {
+    mode: InsertValuesMode.Interval,
+    value: null,
+    isRefField: IS_TIME_FIELD,
+    guessInterval: false,
+  },
+  operator: ({ mode, value = null, isRefField = IS_TIME_FIELD, guessInterval = false, threshold = THRESHOLD_HOUR }) => (
+    source
+  ) =>
+    source.pipe(
+      map((dataFrames) => {
+        if (mode === InsertValuesMode.Interval) {
+          return dataFrames.map((f) => insertValuesInterval(f, value ?? null, guessInterval ?? false, isRefField));
+        } else {
+          return dataFrames.map((f) => insertValuesThreshold(f, value ?? null, threshold, isRefField ?? IS_TIME_FIELD));
+        }
+      })
+    ),
+};
+
+// expects ascending number values in refField
+function getInterval(refField: Field, guessFromValues = false) {
+  let interval = Math.max(refField.config?.interval ?? 0, 0);
+
+  if (interval === 0 && guessFromValues && refField.values.length > 2) {
+    let refVals = refField.values.toArray();
+
+    interval = Infinity;
+
+    for (let i = 1; i < refVals.length; i++) {
+      interval = Math.min(interval, refVals[i] - refVals[i - 1]);
+    }
+  }
+
+  return Number.isFinite(interval) ? interval : 0;
+}
+
+// assumes some interval steps may be absent, but what remains in the time field:
+// 1. contains only numbers, would pass Number.isFinite(value)
+// 2. conforms to the specified interval
+// 3. is in ascending order
+export function insertValuesInterval(
+  frame: DataFrame,
+  value: any = null,
+  guessFromValues = false,
+  isRefField = IS_TIME_FIELD
+) {
+  if (frame.length <= 1) {
+    return frame;
+  }
+
+  let refField = frame.fields.find(isRefField);
+
+  if (refField == null) {
+    return frame;
+  }
+
+  let interval = getInterval(refField, guessFromValues);
+
+  if (interval === 0) {
+    return frame;
+  }
+
+  let refValues = refField.values.toArray();
+  let len = refValues.length;
+  let minValue = refValues[0];
+  let maxValue = refValues[len - 1];
+  let filledLen = (maxValue - minValue) / interval + 1;
+
+  // cheap exit length matches expected length
+  if (len === filledLen) {
+    return frame;
+  }
+
+  let filledFieldValues: any[][] = [];
+
+  for (let field of frame.fields) {
+    let filledValues = Array(filledLen);
+
+    if (field === refField) {
+      for (let i = 0; i < filledLen; i++) {
+        filledValues[i] = minValue + i * interval;
+      }
+    } else {
+      let fieldValues = field.values.toArray();
+
+      for (let i = 0, j = 0; i < filledLen; i++) {
+        filledValues[i] = refValues[j] === minValue + i * interval ? fieldValues[j++] : value;
```

Comment:
`refValues[j] === minValue + i * interval`

there is a bit of risk here in not matching up exact expected interval values when dealing with floats and accumulated precision drift. but since we only deal with millisecond (int) timestamps, this should work okay.

this strategy would be problematic is a minValue is 100.33333 and interval is 0.001. if this becomes an issue it's possible to do equality tests within +/- some precision (e.g. interval / 10), but with significant performance tradeoff.

@gabor @toddtreece is it possible for prometheus to return float time values or values that are off-by-one-ms or similar from the declared interval step?

Reply:
here's an example matrix response from prometheus:

```
{
            "metric" : {
               "__name__" : "up",
               "job" : "prometheus",
               "instance" : "localhost:9090"
            },
            "values" : [
               [ 1435781430.781, "1" ],
               [ 1435781445.781, "1" ],
               [ 1435781460.781, "1" ]
            ]
         },
```

Reply:
that might be the raw response? but i think the backend converts everything to milliseconds for the frontend. what comes over the wire from the prometheus datasource is all in ms timestamps.

i loaded up https://play.grafana.org/d/000000029/prometheus-demo-dashboard?orgId=1&editPanel=1&tab=transform and added a "Filter by name" transformation. the query response in the network tab and the data structure in the debugger all have ms int timestamps, not floats.

can you point me to the steps needed to trigger float UNIX timestamps in a transformer input?

![image](https://user-images.githubusercontent.com/43234/151685831-4c4617e8-42fd-4d2a-b232-2842789dc4e4.png)

![image](https://user-images.githubusercontent.com/43234/151685837-9090b327-e303-4d92-ab5d-7d1c244d7acc.png)

Reply:
@leeoniya yeah, that's correct. they are converted to ints on the backend. sorry, i wasn't sure if you meant the raw response or not